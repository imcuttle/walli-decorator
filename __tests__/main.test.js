/**
 * @file main
 * @author imcuttle
 * @date 2018/4/4
 */
const { default: walli, check } = require('../src')
const w = require('walli')
const decorate = require('decorate-properties')

it('should normal passed', function() {
  class Person {
    @walli(w.string) name = 'imcuttle'
    @walli(w.number) age = 19
    @walli(w.oneOf(['M', 'F']))
    gender = 'X'
  }

  const person = new Person()
  expect(check(person)).toEqual({ gender: "expected: oneOf(['M', 'F']), actual: 'X'." })

  person.age = 's123'
  expect(check(person)).toEqual({
    age: 'expected type: number, actual type: string.',
    gender: "expected: oneOf(['M', 'F']), actual: 'X'."
  })

  class MyPerson extends Person {
    @walli(w.string) age = 'abc'
  }
  // without side effect
  expect(check(person)).toEqual({
    age: 'expected type: number, actual type: string.',
    gender: "expected: oneOf(['M', 'F']), actual: 'X'."
  })

  expect(check(new MyPerson())).toEqual({
    gender: "expected: oneOf(['M', 'F']), actual: 'X'."
  })
})

it('should be executed recursively', function () {

  class Person {
    @walli(w.string) name = 'imcuttle'
    @walli(w.number) age = 19
    @walli(w.oneOf(['M', 'F']))w
    gender = 'X'
  }

  class Family {
    son = new Person()
  }
  // TODO
  console.log(check(new Family()))
})

