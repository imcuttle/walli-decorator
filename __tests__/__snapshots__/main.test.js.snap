// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check: returnWalliResult 1`] = `
Object {
  "abc": UnlawfulnessList {
    "list": Array [
      Unlawfulness {
        "paths": Array [],
        "reasons": Array [
          ToEqualReason {
            "actual": "",
            "expect": OneOf {
              "isRequired": true,
              "msg": null,
              "options": Object {},
              "rule": Array [
                "a",
                "b",
                "c",
              ],
            },
            "msg": null,
            "prefix": "",
            "suffix": "",
          },
        ],
      },
    ],
  },
  "number": UnlawfulnessList {
    "list": Array [
      Unlawfulness {
        "paths": Array [],
        "reasons": Array [
          TypeReason {
            "actual": TypeItem {
              "name": "string",
              "type": [Function],
            },
            "expect": TypeItem {
              "name": "StrictNumber",
              "type": [Function],
            },
            "msg": null,
            "prefix": "",
            "suffix": "",
          },
        ],
      },
    ],
  },
}
`;
